File: cpp_solutions/2816.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* doubleIt(ListNode* head) {
        vector<int> numbers;
        ListNode* curr = head;
        while (curr) {
            numbers.push_back(curr->val);
            curr = curr->next;
        }
        int n = numbers.size();
        int leading = 0;
        for (int i = n - 1; i >= 0; --i) {
            int tmp = numbers[i] * 2 + leading;
            numbers[i] = (tmp) % 10;
            leading = tmp / 10;
        }
        curr = head;
        for (int i = 0; i < n; ++i) {
            curr->val = numbers[i];
            curr = curr->next;
        }
        if (leading) {
            ListNode* extra = new ListNode(leading, head);
            return extra;
        }
        return head;
    }
};

File: rust_solutions/2816.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn double_it(mut head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut st = None;

        while let Some(node) = &mut head {
            std::mem::swap(&mut st, &mut node.next);
            std::mem::swap(&mut st, &mut head);
        }

        let mut x = 0;
        
        while let Some(node) = &mut st {
            x += node.val * 2;
            node.val = x % 10;
            x /= 10;
            std::mem::swap(&mut head, &mut node.next);
            std::mem::swap(&mut head, &mut st);
        }

        if x > 0 {
            head = Some(Box::new(ListNode {val: x, next: head}));
        }

        head
    }
}
